### ISSUE TITLE
QUnit Integration and asynctest not working with this.stub

### ISSUE THREAD

**Issue by Silwing** [sinonjs/sinon#438]
I have the following code:

```javascript
myAjaxFunc(param, callback) {
    jQuery.ajax("url" + param)
    .fail(function (data) {
        alert("Error message.");
        callback([]);
    });
} 

asyncTest("Run function using ajax resulting in error", function () {
    var errorMsg = "";
    var stub = this.stub(window, "alert", function (msg) { errorMsg = msg; return false; });
    myAjaxFunc("Parameter resulting in error", function (data) {
        equal("Error message.", errorMsg, "Expected an error message to be alerted.");
        start();
    });
});
```

However the stubbing does not work inside the ajax fail-callback function. I suspect it has to do with when Sinon restores the stubbed objects in the QUnit workflow. It seems to be done before the asyncTest is completed.

For now I have worked around it by replacing this.sub with sinon.stub and manually calling stub.restore(); just before my call to start();.

**Comment by ThomasChadzelek** [comment]
We basically face the same issue.

http://api.qunitjs.com/QUnit.test/ says that "QUnit.test() can automatically handle the asynchronous resolution of a Promise on your behalf if you return a thenable Promise as the result of your callback function."

This is very elegant, but it does not properly interoperate with sinon.test(callback). The sandbox is restored synchronously when the callback returns. It would be great to wait until the promise returned by the callback has been settled. Any chance that this is added as an option to Sinon.JS?

**Comment by fatso83** [comment]
@ThomasChadzelek : this is what [this PR for sinon-test](https://github.com/sinonjs/sinon-test/pull/7) fixes for `sinon-test` (which uses the sanboxes). Maybe you could see if this could be fixed in the sandbox layer of sinon by copying some of the logic?

**Comment by fatso83** [comment]
With the fix for sinonjs/sinon-test#6 this can be handled by just using `sinon-test` to wrap your functions. Returning a promise will then make sure that all spies and stubs are cleaned up after the promise resolves. 

This of course depends on using the `sinon@next` from NPM along with `sinon-test`.

Does that solve your issues, @ThomasChadzelek and @Silwing?

**Comment by Silwing** [comment]
I haven't worked on the project I used Sinon for in a long while, so I can unfortunately not report back on whether that fix works. By the looks of it it should solve the problem though.

**Comment by ThomasChadzelek** [comment]
@fatso83: Thanks a lot, that fix is basically the same as my code and indeed solves my initial issue. I made some further tests and found out that the issue is actually bigger, though. QUnit offers a "beforeEach" and "afterEach" concept for setup/teardown and you can share spies, stubs, mocks across all tests. The current "sinon-qunit.js" bridge we are using achieves auto-sandboxing, so to say, via sinon.test() only for the test itself, not for setup/teardown. It would be nice if we could integrate that as well and use code like "this.mock" already during setup and can even avoid the teardown for it because it would happen automatically.

In other words: a QUnit test starts with "beforeEach", not just with a call to the test callback itself. And it ends with "afterEach", not just when the test callback returns and not even when the promise is settled. How could this play together with sinon.test()'s automagic sandbox, which I like very much?

**Comment by fatso83** [comment]
@ThomasChadzelek : [testCase](http://sinonjs.org/docs/#sinon-testCase) allows you to share state for tests, including setup and teardown methods.

**Comment by ThomasChadzelek** [comment]
@fatso83: I saw that method mentioned, but frankly, the documentation is way too short to be understandable. I cannot tell how this integrates with QUnit 2.x From the SinonJS code we are running, it seems that /^(setUp|tearDown)$/ is hardcoded, which does not fit to beforeEach/afterEach...

"Test cases

If you need the behavior of `sinon.test` for more than one test method in a test case, you can use `sinon.testCase`, which behaves exactly like wrapping each test in `sinon.test` with one exception: `setUp` and `tearDown` can share fakes.
var obj = sinon.testCase({});"

**Comment by fatso83** [comment]
All it does is automatically create a sandbox and then clean it up again
after the tests have finished.

Two suggestions
1. Submit a pull request to sinon-test with the fix that adds other names
for setup and teardown (or make it just configurable). You already seem to
have done the hard work of finding the relevant code and we have just
recruited new maintainers for the project so this should be smooth sailing.
2. Just manually create a sandbox in `before` and clean it up in `after`.
This is the quick and easy route.
